### PR TITLE
Update to support PHP 8.1

### PR DESCRIPTION
**Pull request by pascalheidmann**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

With this PR [`laminas/laminas-serializer`](https://github.com/laminas/laminas-serializer) runs under PHP 8.1.
It updates several dependencies to allow compatibility especially with [`laminas/laminas-coding-style`](https://github.com/laminas/laminas-coding-style). This results in quite a few changes code style improvements and better code quality.

~As this PR also adds return types & type hints to several classes this might be considered as a breaking change. Those changes are also listed in the newly added `CHANGELOG.md`.~
BCs like return types & type hints for library sources are not part anymore (I left those in tests as those shouldn't be a concern of changes of signatures).

### CODE REVIEW

File: src/AbstractSerializer.php [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+declare(strict_types=1);
+
 namespace Laminas\Serializer;
 
 use Laminas\Serializer\Adapter\AdapterInterface as Adapter;
 use Laminas\ServiceManager\ServiceManager;
+use Traversable;
 
-abstract class Serializer
+abstract class AbstractSerializer
```
Comment (Ocramius) [inline src/AbstractSerializer.php]: This is a major BC break: we can't really introduce this for a dependency upgrade :thinking:

File: src/Adapter/JsonOptions.php [hunk 2]
```diff
@@ -23,63 +23,51 @@ class JsonOptions extends AdapterOptions
     protected $objectDecodeType = LaminasJson::TYPE_ARRAY;
 
     /**
-     * @param  bool $flag
      * @return JsonOptions
      */
-    public function setCycleCheck($flag)
+    public function setCycleCheck(bool $flag): self
```
Comment (Ocramius) [inline src/Adapter/JsonOptions.php]: We can't add return types on non-`final` classes, as it is a BC break

Comment (Ocramius) [reply src/Adapter/JsonOptions.php]: Adding parameter type declarations is a bit milder as a BC break, but still a major problem: before, we would use a type cast, and now users get a crash.

Can't be done here.

File: src/Adapter/PythonPickle.php [hunk 3]
```diff
@@ -1,145 +1,202 @@
...
-    const OP_OBJ             = 'o';     // build & push class instance
-    const OP_PUT             = 'p';     // store stack top in memo; index is string arg
-    const OP_BINPUT          = 'q';     //   "     "    "   "   " ;   "    " 1-byte arg
-    const OP_LONG_BINPUT     = 'r';     //   "     "    "   "   " ;   "    " 4-byte arg
-    const OP_SETITEM         = 's';     // add key+value pair to dict
-    const OP_TUPLE           = 't';     // build tuple from topmost stack items
-    const OP_EMPTY_TUPLE     = ')';     // push empty tuple
-    const OP_SETITEMS        = 'u';     // modify dict by adding topmost key+value pairs
-    const OP_BINFLOAT        = 'G';     // push float; arg is 8-byte float encoding
+    private const OP_MARK            = '(';     // push special markobject on stack
```
Comment (Ocramius) [inline src/Adapter/PythonPickle.php]: These constants were implicitly `public` before - we can't make them `private` here, as it is a BC break.

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by froschdesign** [comment]
@pascalheidmann 
Thank you for your contribution!

[This package is security-only maintenance mode](https://github.com/laminas/laminas-serializer#readme) therefore the support for PHP 8.1 is good, but any breaking changes must be omitted.
Thanks in advance! 👍

File: composer.lock [hunk 4]
```diff
@@ -1,2657 +0,0 @@
-{
```
Comment (Ocramius) [inline composer.lock]: I see `composer.lock` was completely deleted: we should really have it at its original state, if possible.

See how upgrades were done elsewhere, such as in https://github.com/laminas/laminas-diactoros/pull/70, which uses a `.laminas-ci.json` to make CI run also when deps are locked

File: src/Adapter/PhpSerialize.php [hunk 5]
```diff
@@ -97,14 +106,14 @@ public function serialize($value)
...
-     * @throws Exception\RuntimeException on unserialize error
+     * @throws Exception\RuntimeException On unserialize error.
      */
     public function unserialize($serialized)
     {
         if (! is_string($serialized) || ! preg_match('/^((s|i|d|b|a|O|C):|N;)/', $serialized)) {
             $value = $serialized;
             if (is_object($value)) {
-                $value = get_class($value);
+                $value = $value::class;
```
Comment (Ocramius) [inline src/Adapter/PhpSerialize.php]: `$value::class` won't work on PHP 7.3, I think?

File: src/Adapter/PythonPickle.php [hunk 6]
```diff
@@ -737,33 +815,35 @@ protected function loadBinPut()
...
+     * @return void
      */
     protected function loadLongBinPut()
     {
         $bin = $this->read(4);
         if (static::$isLittleEndian === false) {
             $bin = strrev($bin);
         }
-        list(, $id) = unpack('l', $bin);
+        [, $id] = unpack('l', $bin);
```
Comment (Ocramius) [inline src/Adapter/PythonPickle.php:833]: Does the `[, $id] = ` syntax also work on 7.3?

File: src/AdapterPluginManager.php [hunk 7]
```diff
@@ -96,9 +110,9 @@ public function validate($instance)
         if (! $instance instanceof $this->instanceOf) {
             throw new InvalidServiceException(sprintf(
                 '%s can only create instances of %s; %s is invalid',
-                get_class($this),
+                static::class,
                 $this->instanceOf,
-                (is_object($instance) ? get_class($instance) : gettype($instance))
+                is_object($instance) ? $instance::class : gettype($instance)
```
Comment (Ocramius) [inline src/AdapterPluginManager.php]: Similar to above, `::class` is fairly new on variables

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Adapter/PhpSerialize.php (hunk 5 above)
Comment (driehle) [reply src/Adapter/PhpSerialize.php]: That was introduced in PHP 8.0 (https://3v4l.org/qVnSI).

File: src/Adapter/PythonPickle.php (hunk 6 above)
Comment (driehle) [reply src/Adapter/PythonPickle.php:833]: This should be fine, seems it has been introduced in PHP 7.1 (https://3v4l.org/8qrE6).

File: src/AdapterPluginManager.php [hunk 8]
```diff
@@ -43,24 +56,25 @@ class AdapterPluginManager extends AbstractPluginManager
...
-        \Zend\Serializer\Adapter\PhpSerialize::class => Adapter\PhpSerialize::class,
-        \Zend\Serializer\Adapter\PythonPickle::class => Adapter\PythonPickle::class,
-        \Zend\Serializer\Adapter\Wddx::class => Adapter\Wddx::class,
+        igbinary::class     => Adapter\IgBinary::class,
+        json::class         => Adapter\Json::class,
+        msgpack::class      => Adapter\MsgPack::class,
+        phpcode::class      => Adapter\PhpCode::class,
+        phpserialize::class => Adapter\PhpSerialize::class,
+        pythonpickle::class => Adapter\PythonPickle::class,
+        wddx::class         => Adapter\Wddx::class,
```
Comment (Ocramius) [inline src/AdapterPluginManager.php]: Casing seems to have broken here? 🤔

Comment (pascalheidmann) [reply src/AdapterPluginManager.php]: Well, looks like you are right and `phpcs` has some bug there (auto-import happened by `phpcs`/`phpbf`) :/
Will fix it & have also a look at `phpcs`'s fix for this behavior 🤔

**Review by Ocramius — APPROVED** [review]
LGTM, thanks @pascalheidmann!

**Pull request merged**